File: solution/python/12789.py
N = input()
num_list = list(map(int,input().split()))

stack = []
number = 1
ans = "Nice\n"
for num in num_list:
    if num == number:
        number += 1
    else:
        while len(stack)!=0 and stack[-1]==number:
            number = stack.pop() + 1
        if (len(stack)!=0 and num<stack[-1]) or len(stack)==0:
            stack.append(num)
        else:
            ans = "Sad\n"
            break
print(ans)

File: solution/cpp/12789.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>
#include <map>
#include <string>
#include <iostream>
#include <stack>

using namespace std;

int main(void) {
    int N; // number of instructions
    int number = 1;
    
    string ans = "Nice\n";
    stack<int> stack;
    
    scanf("%d",&N);
    
    for(int i=0;i<N;i++) {
        int temp;
        scanf("%d",&temp);

        if (temp==number) number++;
        else {
            while(stack.size()!=0 && stack.top()==number) {
                number = stack.top() + 1;
                stack.pop();
            }

            if (stack.size()==0) stack.push(temp);
            else if (stack.size()!=0 && temp<stack.top()) stack.push(temp);
            else {
                ans = "Sad\n";
                break;
            }
        }
    }
    cout << ans;
}
